File: scala/luhn/src/main/scala/Luhn.scala
object Luhn {
  def apply(value: Long) = new Luhn(value)
}

class Luhn(val value: Long) {
  lazy val checkDigit = value % 10

  lazy val addends =
    Seq.iterate(value, math.log10(value).toInt + 1)(_ / 10)
      .map(_ % 10)
      .zipWithIndex
      .map{ case (d, i) => if(i%2==1) d*2 else d }
      .map(d=> if(d > 9) d - 9 else d)
      .reverse

  lazy val checksum = addends.sum % 10

  lazy val isValid = checksum == 0

  lazy val create =
    (0 to 9)
      .map(value*10 + _)
      .map(Luhn.apply)
      .find(_.isValid)
        .get
          .value
}

File: java/luhn/src/main/java/Luhn.java
import java.util.Arrays;
import java.util.stream.IntStream;

import static java.util.Arrays.stream;

public class Luhn {
    private final int number;

    public Luhn(int number) {
        this.number = number;
    }

    public int getCheckDigit() {
        return number%10;
    }

    public int[] getAddends() {
        final int[] digits = toDigits(number);

        final int len = digits.length;

        return doubleEverySecond(digits, len%2);
    }

    public int getCheckSum() {
        return stream(getAddends()).sum();
    }

    public boolean isValid() {
        return getCheckSum()%10 == 0;
    }

    public static int[] toDigits(int number) {
        final int[] reversed =
                IntStream.iterate(number, nr-> nr / 10)
                        .limit((long)Math.log10(number) + 1)
                        .map(nr-> nr % 10)
                        .toArray();

        return reverseArray(reversed);
    }

    public static int[] doubleEverySecond(int[] is, int evenOdd) {
        return IntStream.range(0, is.length)
                .map(i-> i%2==evenOdd? is[i]*2: is[i])
                .map(i-> i > 9? i-9: i)
                .toArray();
    }

    public static long create(int number) {
        final int[] digits = toDigits(number);
        final int len = digits.length;

        final int[] doubled = doubleEverySecond(digits, 1 - len%2);

        final long checksum = (10 - (Arrays.stream(doubled).sum() % 10)) % 10;

        return (long)number * 10 + checksum;
    }

    public static int checksum(int[] is) {
        return stream(is).sum();
    }

    public static int[] reverseArray(int[] array) {
        final int len = array.length;
        return IntStream.rangeClosed(1, len)
                .map(i-> array[len-i])
                .toArray();
    }
}
